File: forge-dropship/scripts/get-shopify-token.mjs
#!/usr/bin/env node
// get-shopify-token.mjs — mint a durable Shopify Admin API token for Everaly.
//
// Node twin of get_shopify_token.py, for machines with Node but no Python
// (the Windows workstation). Same flow, same output, no dependencies.
//
// Why this exists: Everaly has no legacy "custom app" flow, so there is no
// `shpat_` token to reveal in the admin. A Dev Dashboard app exposes only a
// Client ID, a Secret, and a CI/CD automation token (`atkn_`) — none of which
// authenticate against /admin/api. The supported path is a one-time OAuth
// authorization-code exchange, which returns an OFFLINE token that does not
// expire. That is what SHOPIFY_ADMIN_TOKEN wants.
//
// The secret comes from an env var or a hidden prompt, never from argv. The
// token is written straight into dropship.env and never printed.
//
// --- Before running, in the Dev Dashboard for your app ---
//   1. Configuration -> Admin API access scopes, tick:
//        read_products, read_inventory, read_orders,
//        read_fulfillments, read_customers, read_locations
//   2. Configuration -> Redirect URLs, add exactly:
//        http://localhost:3456/callback
//   3. Release a new version so the config goes live.
//   4. Settings -> copy the Client ID, reveal the Secret.
//
// --- Run ---
//   node forge-dropship/scripts/get-shopify-token.mjs --client-id <CLIENT_ID>

import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { fileURLToPath } from 'node:url';
import { spawn } from 'node:child_process';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ENV_PATH = path.resolve(HERE, '..', 'config', 'dropship.env');

const PORT = 3456;
const REDIRECT_URI = `http://localhost:${PORT}/callback`;
const SCOPES = [
  'read_products',
  'read_inventory',
  'read_orders',
  'read_fulfillments',
  'read_customers',
  'read_locations',
].join(',');

function arg(name) {
  const i = process.argv.indexOf(`--${name}`);
  return i > -1 ? process.argv[i + 1] : '';
}

function readEnv(file) {
  const cfg = {};
  if (!fs.existsSync(file)) return cfg;
  for (const raw of fs.readFileSync(file, 'utf8').split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith('#') || !line.includes('=')) continue;
    const i = line.indexOf('=');
    cfg[line.slice(0, i).trim()] = line.slice(i + 1).trim();
  }
  return cfg;
}

// Replace the SHOPIFY_ADMIN_TOKEN line in place, leaving everything else alone.
function writeToken(file, token) {
  const original = fs.existsSync(file) ? fs.readFileSync(file, 'utf8') : '';
  const line = `SHOPIFY_ADMIN_TOKEN=${token}`;
  const re = /^[ \t]*SHOPIFY_ADMIN_TOKEN[ \t]*=.*$/m;
  const updated = re.test(original)
    ? original.replace(re, line)
    : original.replace(/\n*$/, '\n') + line + '\n';
  fs.writeFileSync(file, updated, 'utf8');
}

function promptHidden(question) {
  return new Promise((resolve) => {
    process.stdout.write(question);
    const stdin = process.stdin;
    const wasRaw = stdin.isRaw;
    stdin.setRawMode?.(true);
    stdin.resume();
    stdin.setEncoding('utf8');
    let buf = '';
    const onData = (ch) => {
      if (ch === '\r' || ch === '\n' || ch === '') {
        stdin.setRawMode?.(wasRaw ?? false);
        stdin.pause();
        stdin.removeListener('data', onData);
        process.stdout.write('\n');
        resolve(buf);
      } else if (ch === '') {
        process.exit(130);
      } else if (ch === '' || ch === '\b') {
        buf = buf.slice(0, -1);
      } else {
        buf += ch;
      }
    };
    stdin.on('data', onData);
  });
}

function openBrowser(url) {
  const cmd = process.platform === 'win32' ? 'cmd'
    : process.platform === 'darwin' ? 'open' : 'xdg-open';
  const args = process.platform === 'win32' ? ['/c', 'start', '', url] : [url];
  try { spawn(cmd, args, { detached: true, stdio: 'ignore' }).unref(); } catch { /* user can paste it */ }
}

const die = (msg) => { console.error(msg); process.exit(1); };

const clientId = arg('client-id');
if (!clientId || process.argv.includes('--help')) {
  console.log('usage: node get-shopify-token.mjs --client-id <CLIENT_ID> [--shop <store>.myshopify.com]');
  process.exit(clientId ? 0 : 1);
}

const cfg = readEnv(ENV_PATH);
const shop = arg('shop') || cfg.SHOPIFY_STORE_DOMAIN || '';
if (!shop) die('No shop domain. Pass --shop <store>.myshopify.com');
if (!shop.endsWith('.myshopify.com')) die(`Shop must be the myshopify domain, got "${shop}"`);

let secret = process.env.SHOPIFY_CLIENT_SECRET || '';
if (!secret) secret = (await promptHidden("Paste the app's Client Secret (input hidden): ")).trim();
if (!secret) die('No client secret given.');

const state = crypto.randomBytes(24).toString('base64url');
const authUrl = `https://${shop}/admin/oauth/authorize?` + new URLSearchParams({
  client_id: clientId,
  scope: SCOPES,
  redirect_uri: REDIRECT_URI,
  state,
}).toString();

console.log(`\nStore : ${shop}`);
console.log(`Scopes: ${SCOPES}\n`);
console.log('Opening your browser. Approve the install there.');
console.log(`If it does not open, paste this in manually:\n\n${authUrl}\n`);

const cb = await new Promise((resolve) => {
  const server = http.createServer((req, res) => {
    const u = new URL(req.url, `http://localhost:${PORT}`);
    if (u.pathname !== '/callback') { res.writeHead(404).end(); return; }
    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end("<html><body style='font:16px system-ui;padding:40px'><h2>Done.</h2>"
      + '<p>Token captured. Close this tab and go back to the terminal.</p></body></html>');
    resolve({ code: u.searchParams.get('code') || '', state: u.searchParams.get('state') || '' });
    setTimeout(() => server.close(), 100);
  });
  server.listen(PORT, 'localhost', () => openBrowser(authUrl));
});

if (!cb.code) die('No authorization code came back.');
if (cb.state !== state) die('State mismatch — aborting rather than trusting that callback.');

const res = await fetch(`https://${shop}/admin/oauth/access_token`, {
  method: 'POST',
  headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
  body: JSON.stringify({ client_id: clientId, client_secret: secret, code: cb.code }),
});
if (!res.ok) die(`Token exchange failed: HTTP ${res.status} ${(await res.text()).slice(0, 300)}`);

const payload = await res.json();
const token = payload.access_token || '';
if (!token) die(`No access_token in the response: ${Object.keys(payload).join(', ')}`);

writeToken(ENV_PATH, token);
console.log(`\nWrote SHOPIFY_ADMIN_TOKEN to ${ENV_PATH}`);
console.log(`  length ${token.length}, prefix ${token.slice(0, 6)}…`);
console.log(`  granted scopes: ${payload.scope || '?'}`);
console.log('\nThe token was not printed. Copy the same line to the box at');
console.log('  /opt/forge/forge-dropship/config/dropship.env');
console.log('then: systemctl restart forge-reios');

File: forge-dropship/scripts/get_shopify_token.py
#!/usr/bin/env python3
"""get_shopify_token.py — mint a durable Shopify Admin API token for Everaly.

Why this exists: this store has no legacy "custom app" flow, so there is no
`shpat_` token to reveal in the admin. Dev Dashboard apps expose only a Client ID,
a Secret, and a CI/CD automation token (`atkn_`) — none of which authenticate
against /admin/api. The supported path is a one-time OAuth authorization-code
exchange, which returns an OFFLINE access token that does not expire. That token
is what `SHOPIFY_ADMIN_TOKEN` wants.

The secret is read from an env var or a hidden prompt, never from argv (argv shows
up in shell history and process lists). The resulting token is written straight
into dropship.env and is never printed.

Stdlib only, to match dropship_env.py.

--- Before running, in the Dev Dashboard (dev.shopify.com) for your app ---

  1. Configuration -> Admin API access scopes, tick:
       read_products, read_inventory, read_orders,
       read_fulfillments, read_customers, read_locations
  2. Configuration -> Redirect URLs, add exactly:
       http://localhost:3456/callback
  3. Release a new version so the config is live.
  4. Settings -> copy the Client ID, and reveal the Secret.

--- Run ---

    cd forge-dropship/scripts
    python get_shopify_token.py --client-id <CLIENT_ID>

  It opens your browser, you click Install, and it writes SHOPIFY_ADMIN_TOKEN
  into ../config/dropship.env. Re-running overwrites the line.
"""
from __future__ import annotations

import argparse
import getpass
import http.server
import json
import os
import re
import secrets
import sys
import threading
import urllib.parse
import urllib.request
import webbrowser
from pathlib import Path

HERE = Path(__file__).resolve().parent
ENV_PATH = HERE.parent / "config" / "dropship.env"

PORT = 3456
REDIRECT_URI = f"http://localhost:{PORT}/callback"

SCOPES = ",".join([
    "read_products",
    "read_inventory",
    "read_orders",
    "read_fulfillments",
    "read_customers",
    "read_locations",
])

_result: dict = {}


def read_env(path: Path) -> dict:
    cfg = {}
    if path.exists():
        for line in path.read_text(encoding="utf-8").splitlines():
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                k, v = line.split("=", 1)
                cfg[k.strip()] = v.strip()
    return cfg


def write_token(path: Path, token: str) -> None:
    """Replace the SHOPIFY_ADMIN_TOKEN line in place, preserving everything else."""
    original = path.read_text(encoding="utf-8") if path.exists() else ""
    line = f"SHOPIFY_ADMIN_TOKEN={token}"
    if re.search(r"(?m)^[ \t]*SHOPIFY_ADMIN_TOKEN[ \t]*=.*$", original):
        updated = re.sub(r"(?m)^[ \t]*SHOPIFY_ADMIN_TOKEN[ \t]*=.*$", line, original, count=1)
    else:
        updated = original.rstrip("\n") + "\n" + line + "\n"
    path.write_text(updated, encoding="utf-8")


class Handler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):  # noqa: N802
        parsed = urllib.parse.urlparse(self.path)
        if parsed.path != "/callback":
            self.send_response(404)
            self.end_headers()
            return
        q = urllib.parse.parse_qs(parsed.query)
        _result["code"] = (q.get("code") or [""])[0]
        _result["state"] = (q.get("state") or [""])[0]
        _result["shop"] = (q.get("shop") or [""])[0]
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.end_headers()
        self.wfile.write(
            b"<html><body style='font:16px system-ui;padding:40px'>"
            b"<h2>Done.</h2><p>Token captured. You can close this tab and go back to the terminal.</p>"
            b"</body></html>"
        )
        threading.Thread(target=self.server.shutdown, daemon=True).start()

    def log_message(self, *args):  # silence access logging
        return


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--client-id", required=True, help="Dev Dashboard app Client ID")
    ap.add_argument("--shop", default="", help="myshopify domain (defaults to the one in dropship.env)")
    args = ap.parse_args()

    cfg = read_env(ENV_PATH)
    shop = args.shop or cfg.get("SHOPIFY_STORE_DOMAIN", "")
    if not shop:
        print("No shop domain. Pass --shop <store>.myshopify.com", file=sys.stderr)
        return 1
    if not shop.endswith(".myshopify.com"):
        print(f"Shop must be the myshopify domain, got {shop!r}", file=sys.stderr)
        return 1

    # Secret from env or hidden prompt. Never from argv.
    secret = os.environ.get("SHOPIFY_CLIENT_SECRET", "")
    if not secret:
        secret = getpass.getpass("Paste the app's Client Secret (input hidden): ").strip()
    if not secret:
        print("No client secret given.", file=sys.stderr)
        return 1

    state = secrets.token_urlsafe(24)
    auth_url = (
        f"https://{shop}/admin/oauth/authorize?"
        + urllib.parse.urlencode({
            "client_id": args.client_id,
            "scope": SCOPES,
            "redirect_uri": REDIRECT_URI,
            "state": state,
            # offline == a token that does not expire, which is what a daemon needs
            "grant_options[]": "",
        })
    )

    print(f"\nStore : {shop}")
    print(f"Scopes: {SCOPES}\n")
    print("Opening your browser. Approve the install there.")
    print(f"If it does not open, paste this in manually:\n\n{auth_url}\n")

    server = http.server.HTTPServer(("localhost", PORT), Handler)
    threading.Thread(target=webbrowser.open, args=(auth_url,), daemon=True).start()
    server.serve_forever()  # Handler shuts this down once the callback lands

    if not _result.get("code"):
        print("No authorization code came back.", file=sys.stderr)
        return 1
    if _result.get("state") != state:
        print("State mismatch — aborting rather than trusting that callback.", file=sys.stderr)
        return 1

    body = json.dumps({
        "client_id": args.client_id,
        "client_secret": secret,
        "code": _result["code"],
    }).encode()
    req = urllib.request.Request(
        f"https://{shop}/admin/oauth/access_token",
        data=body,
        headers={"Content-Type": "application/json", "Accept": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as r:
            payload = json.loads(r.read().decode())
    except urllib.error.HTTPError as e:
        print(f"Token exchange failed: HTTP {e.code} {e.read().decode()[:300]}", file=sys.stderr)
        return 1

    token = payload.get("access_token", "")
    if not token:
        print(f"No access_token in the response: {list(payload)}", file=sys.stderr)
        return 1

    write_token(ENV_PATH, token)
    print(f"\nWrote SHOPIFY_ADMIN_TOKEN to {ENV_PATH}")
    print(f"  length {len(token)}, prefix {token[:6]}…")
    print(f"  granted scopes: {payload.get('scope', '?')}")
    print("\nThe token was not printed. Copy the same line to the box at")
    print("  /opt/forge/forge-dropship/config/dropship.env")
    print("then: systemctl restart forge-reios")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
